[스택, 큐, 덱] 9월 15일

jaeseo222 opened the pull request.

Commit:
[스택, 큐, 덱] 9월 15일

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/1918.cpp:
```diff
@@ -0,0 +1,48 @@
+		}
+
+		switch (str[i])
+		{
+		case '(': s.push('('); break;
```
일관성을 유지하려면 s.push(str[i])이 좋아요!

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/1918.cpp:
```diff
@@ -0,0 +1,48 @@
+#include<stack>
+using namespace std;
+
+bool check(char c1, char c2) {
+	//-,+면 '(' 나올 때까지만 pop하면 되니까 true
+	//*,/ 는 다음 *,/ 오면 또 pop해야 되니까 true
+	if (c1 == '-' || c1 == '+' || (c1 == '*' || c1 == '/') && (c2 == '*' || c2 == '/')) return true;
+	else return false;
+}
```
오 재미있는 함수를 사용하셨네요!
근데 이렇게 봐서는 조건문이 뭘 의미하는지 알기 힘들어요. 연산자의 우선순위를 리턴하는 함수로 바꿔보는건 어떨까요?

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/1935.cpp:
```diff
@@ -0,0 +1,43 @@
+
+	for (int i = 0; i < str.size(); i++)
+	{
+		if (str[i] >= 'A' && str[i] <= 'Z') {
+			s.push(alph[str[i] - 'A']); //'A'-'A'= 0
+			continue;
+		}
```
아주 잘 끊으셨습니다!

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/1935.cpp:
```diff
@@ -0,0 +1,43 @@
+		double num2 = s.top();
+		s.pop();
+
+		switch (str[i])
+		{
+		case '+': s.push(num2 + num1); break;
+		case '-': s.push(num2 - num1); break;
+		case '*': s.push(num2 * num1); break;
+		case '/': s.push(num2 / num1);
+		}
```
이 부분을 연산결과를 리턴하는 함수로 만들어도 좋아요~

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/2019 카카오 개발자 겨울 인턴십.cpp:
```diff
@@ -0,0 +1,30 @@
+	int answer = 0;
+	stack<int> s; //바구니
+
+	for (int i = 0; i < moves.size(); i++)
+	{
+		int check = moves[i] - 1;
+		for (int j = 0; j < board.size(); j++)
+		{
+			if (board[j][check] != 0) {
+				if (!s.empty() && s.top() == board[j][check])
```
인덴테이션이 너무 깊게 들어가네요...!
열마다의 인형 상태를 스택에 넣는 전처리를 하면 어떨까요?

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/20920.cpp:
```diff
@@ -0,0 +1,32 @@
+		if (m <= word.size()) cnt[word]++;
+	}
+
+	vector<pair<string, int>> voca(cnt.begin(), cnt.end());
+	sort(voca.begin(), voca.end(), cmp);// 정렬된 단어장
```
맵을 바로 벡터에 넣으셨네요! 좋은 전략입니다

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/5430.cpp:
```diff
@@ -0,0 +1,75 @@
+#include<string>
+using namespace std;
+
+void dequeAssign(deque<int>& dq, string data) {
+	string tmp = ""; // 십의자리수 넘어가는 숫자들 고려
+	for (int i = 0; i < data.size(); i++) {
+		if (isdigit(data[i])) tmp += data[i];
+		else {  //'[' ']' ','
+			if (tmp == "") continue; //빈문자열 넣으면 에러
+			dq.push_back(stoi(tmp));
+			tmp = "";
+		}
+	}
+}
```
함수로 잘 빼셨어요! 이 함수가 덱을 리턴해도 좋을 것 같아요!

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/5430.cpp:
```diff
@@ -0,0 +1,75 @@
+	}
+}
+
+void operation(deque<int>& dq, string fun, bool& funR, bool& error) {
+	for (int i = 0; i < fun.size(); i++)
+	{
+		if (fun[i] == 'R') funR = !funR; // R 나오면 지금 상태의 역순으로
+		else if (fun[i] == 'D') {
+			if (dq.empty()) { //D인데 비어있으면 에러
+				error = true;
+				break;
+			}
+
+			funR ? dq.pop_back() : dq.pop_front(); // 역순으로 : 원래대로
+		}
+	}
+}
```
여기도 함수화 하신 건 너무 좋지만, void보다는 뭔갈 리턴하는게 좋아요!

HyunJiLim0406 reviewed (commented):
p3
수고하셨습니다! 논리나 시간복잡도면에서는 드릴 조언이 없는 코드들입니다.
함수화 할 수 있는 부분들도 많이 잡아내셨네요!
다만, 가독성면에서 약간 아쉬운 부분들이 사알짝 있어요! 조금만 수정하고 merge할까요?
코드를 다 짜신 뒤에 한번 살펴보며 더 함수화 할 수 있는 곳은 없는지, 비효율적인 부분은 없는지 등등을 살펴보는 것도 좋아요!
수정하시고 리뷰어로 저 호출해주세요~

Commit:
[스택, 큐, 덱] 9월 16일 - ing

1918에서 연산자의 우선순위를 리턴하는 함수로 바꿔보는 걸 제안해주셨는데 이해가 잘 안 가서,,!! 우선 수정했는데 이게 맞는 건지 모르겠습니다!!!
그리고 카카오 문제에서 해결하지 못하고 있습니다..! stack인 col vector에 board 값을 넣어주려고 하는데 괄호들 때문에 for문의 size를 어떻게 맞춰야 되는 지 감이,, 오질 않습니다. 도와주,,시면 감사하겠습니다,,,!

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/1918.cpp:
```diff
@@ -0,0 +1,51 @@
+#include<stack>
+using namespace std;
+
+bool check(char c1, char c2) {
+	//-,+면 '(' 나올 때까지만 pop하면 되니까 true
+	if (c1 == '-' || c1 == '+') return true;
+
+	//*,/ 는 다음 *,/ 오면 또 pop해야 되니까 true
+	if ((c1 == '*' || c1 == '/') && (c2 == '*' || c2 == '/')) return true;
+
+	else return false;
+}
```
힌트를 더 드리자면...! 이 함수가 bool 말고 int를 리턴하도록 하면 어떨까요? 그렇다면 이 함수는 어떤 인자에 대해 어떤 것들을 리턴하게 될까요?

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/1935.cpp:
```diff
@@ -0,0 +1,48 @@
+const int SIZE = 26;
+int alph[SIZE];
+
+double operation(char c, double n1, double n2) {
+	switch (c)
+	{
+	case '+': return n2 + n1;
+	case '-': return n2 - n1;
+	case '*': return n2 * n1;
+	case '/': return n2 / n1;
+	}
+}
```
아주 좋아요!

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/2019 카카오 개발자 겨울 인턴십.cpp:
```diff
@@ -0,0 +1,33 @@
+	stack<int> s; //바구니
+	vector<stack<int>> col(board.size());// board 열
+
+	for (int i = 0; i < board.size(); i++)
+		for (int j = board.size() - 1; j >= 0; j--) {
+			if (board[j][i] != 0 && isdigit(board[j][i])) col[i].push(board[j][i]);
+		}//[]때문에 말썽,,
```
board[j][i]는 항상 숫자 아닐까요?

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/2019 카카오 개발자 겨울 인턴십.cpp:
```diff
@@ -0,0 +1,33 @@
+
+	for (int i = 0; i < moves.size(); i++)
+	{
+		if (col[moves[i] - 1].top() == 0) continue;
+
+		if (!s.empty() && s.top() == col[moves[i] - 1].top())
+		{
+			s.pop();
+			answer += 2;//2개씩 터지니까
+		}
+		else s.push(col[moves[i] - 1].top());
+
+		col[moves[i] - 1].pop();
```
p1
제가 제대로 이해하지 못한걸수도 있는데 이부분 코드가 위험해보여요
위에서 0이 아닌것만 col 스택에 넣었는데 0을 체크하는게 의미가 있을까요? 그리고 col에 대한 empty 체크가 없어보입니다..! 여기 꼭 수정하셔야 해요
혹시 제가 뭔갈 잘못 이해한거라면 말해주세요!!

HyunJiLim0406 commented on 9월 10일 - 스택, 큐, 덱/5430.cpp:
```diff
@@ -0,0 +1,78 @@
+
+		cin >> fun >> n >> data;
+
+		//data의 수들을 덱에 대입
+		dq = dequeAssign(dq, data);
+
+		//명령어 처리 된 덱
+		dq = operation(dq, fun, funR, error);
```
좋아요~

HyunJiLim0406 reviewed (changes requested):
p1
다 잘 수정해주셨습니다! 근데 지금 크레인 문제에서 런타임에러가 발생할 수도 있을 것 같은 부분이 보여요...! 확인해주시고 다시 리뷰어로 저 호출해주세요!!

Commit:
[우선순위 큐] 9월 20일

HyunJiLim0406 commented on 9월 14일 - 우선순위 큐/11000.cpp:
```diff
@@ -0,0 +1,25 @@
+
+	pq.push(time[0].second);//첫 수업 종료시간
+
+	for (int i = 1; i < n; i++)
+	{
+		if (pq.top() <= time[i].first) pq.pop(); //종료시간보다 시작시간이 크면 수업 이어서 할 수 있으니까 pop
+		pq.push(time[i].second);
+	}
+	cout << pq.size();
```
👍

HyunJiLim0406 commented on 9월 14일 - 우선순위 큐/11286.cpp:
```diff
@@ -0,0 +1,20 @@
+	int n, x;
+	cin >> n;
+
+	priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq; //pair로 절댓값, 원래 값 저장, 오름차순 정렬
+
```
p1
핳 사실 이 문제는 정렬의 compare 함수처럼 우선순위 큐에서의 compare 함수를 정의하는 연습을 위한 문제였습니다!
abs(x)를 따로 저장하지 않고 풀어볼까요?

HyunJiLim0406 commented on 9월 14일 - 우선순위 큐/11723.cpp:
```diff
@@ -0,0 +1,38 @@
+int main() {
+	ios::sync_with_stdio(false);
+	cin.tie(NULL);
+
+	bool s[21] = {false}; //20개의 숫자에 체크/언체크
```
잘 푸셨습니다~
근데 이거랑 같은 논리로 배열없이 풀 수 있는 방법이 있어요!
비트마스킹이라는 방법인데요. 이건 좀 어려울 수도 있으니까 알아만 보셔도 좋아요~

HyunJiLim0406 commented on 9월 14일 - 우선순위 큐/3613.cpp:
```diff
@@ -0,0 +1,59 @@
+
+	//1. 첫 문자가 대문자 또는 _
+	//2. 마지막 문자가 _ 일 때 오류
+	if (str[0] >= 'A' && str[0] <= 'Z' || str[0] == '_' || str[size] == '_') {
+		cout << "Error!" << '\n';
+		return 0;
+	}
```
좋아요! 알파벳 체크에서는 isupper 함수를 쓸 수도 있겠네요!

HyunJiLim0406 commented on 9월 14일 - 우선순위 큐/3613.cpp:
```diff
@@ -0,0 +1,59 @@
+using namespace std;
+vector<char> value;
+
+void javaToCpp(string s) {
+	for (int i = 0; i < s.length(); i++)
+	{
+		if (s[i] >= 'A' && s[i] <= 'Z') value.push_back('_');
+		value.push_back(tolower(s[i]));
+	}
+}
+void cppToJava(string s) {
+	for (int i = 0; i < s.length(); i++)
+		if (s[i] == '_') {
+			value.push_back(toupper(s[i + 1]));
+			i++;
+		}
+		else value.push_back(s[i]);
+}
```
얘네가 그냥 string을 리턴하는게 좋아요~!

HyunJiLim0406 commented on 9월 14일 - 우선순위 큐/3613.cpp:
```diff
@@ -0,0 +1,59 @@
+		}
+		else value.push_back(s[i]);
+}
+
+int main() {
```
지금 언어 체크하는 코드도 좋지만 cppToJava나 javaToCpp 함수처럼 함수를 따로 만드는 것도 좋아요~

HyunJiLim0406 commented on 9월 14일 - 우선순위 큐/7662.cpp:
```diff
@@ -0,0 +1,32 @@
+	cin >> t;
+	while (t--)
+	{
+		multiset<int> s;//key 값 중복 set
+
```
잘 사용하셨어요! 멀티셋없이 우선순위 큐로만 푸는 풀이도 고민해보세요~

HyunJiLim0406 reviewed (changes requested):
p1
허허 pr이 섞였네요! 브랜치를 분리하거나 이전 pr을 머지하는 방법으로 겹치지 않게 해주세요!
전반적으로 아주 잘 푸셨어요! 11286번만 살짝 수정해볼까요?

수정 후 리뷰어로 저 호출해주세요~

추석 잘 보내세요 재서님!

Commit:
[우선순위 큐] 9월 22일 - Update

Commit:
[스택, 큐, 덱] 9월 22일 - Update

jaeseo222 commented:
제가 코멘트를 달아도 되는 지 모르겠지만,,, 튜터님 모두 연휴 동안에도 정성 가득한 피드백 해주셔서 너무 감사드려요!!
특히 저의 부족한 코드를.......... 봐주신 ....... 현지님 정말 감사드립니다..! 😭😭
덕분에 새로운 함수도 알게됐고, 튜터님께서 올려주신 (진짜 쩌는.. )풀이를 참고하여 로직을 생각해 볼 수 있었습니다. 핳ㅎ><
그럼 연휴 즐겁게 보내셨길, 그리고 앞으로도 즐거운 시간만 보내시길 바랍니다!!

HyunJiLim0406 reviewed (commented):
p3
하핳 감사합니다...!
부족하다니요! 저희도 코드리뷰를 하면서 다양한 분들의 코드를 볼 수 있고, 그 과정에서 많이 배우고 있어요~
이제 머지하셔도 됩니다! 연휴가 하루남았네요! 마지막 날까지 즐거운 한가위 되세요!
